File: SubscriptionApi/SubscriptionApi/Commands/AssetQueriesFactory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using SubscriptionApi.Authorization;
using SubscriptionApi.Dto;
using SubscriptionApi.Queries;

namespace SubscriptionApi.Commands
{
  public interface IAssetQueriesFactory
  {
    List<AssetQuery> CreateFrom(IEnumerable<AssetRequestDto> requests);
  }

  public class AssetQueriesFactory : IAssetQueriesFactory
  {
    private readonly AssetQueryResolution _authorizationStructure;

    public AssetQueriesFactory(AssetQueryResolution authorizationStructure)
    {
      _authorizationStructure = authorizationStructure;
    }

    public List<AssetQuery> CreateFrom(IEnumerable<AssetRequestDto> requests)
    {
      return requests.Select(
        assetRequestDto => AssetQueryFor(
          assetRequestDto.Name, 
          assetRequestDto.Kind)).ToList();
    }

    private AssetQuery AssetQueryFor(string name, AssetKind assetKind)
    {
      switch (assetKind)
      {
        case AssetKind.User:
          return new UserQuery(name, _authorizationStructure);
        case AssetKind.OrganizationalUnit:
          return new OrganizationalUnitQuery(name, _authorizationStructure);
        case AssetKind.Hardware:
          return new HardwareQuery(name, _authorizationStructure);
        default:
          throw new ArgumentOutOfRangeException();
      }
    }
  }
}

File: Java/factories/_01_encapsulation_of_global_context/src/main/java/com/github/TrainingExamples/grzesiek_galezowski/commands/AssetQueriesFactory.java
package com.github.TrainingExamples.grzesiek_galezowski.commands;

import authorization.AssetQueryResolution;
import dto.AssetKind;
import dto.AssetRequestDto;
import queries.AssetQuery;
import queries.HardwareQuery;
import queries.OrganizationalUnitQuery;
import queries.UserQuery;

import java.util.List;
import java.util.stream.StreamSupport;

import static java.util.stream.Collectors.toList;

public class AssetQueriesFactory implements IAssetQueriesFactory {
    private final AssetQueryResolution authorizationStructure;

    public AssetQueriesFactory(AssetQueryResolution authorizationStructure) {
        this.authorizationStructure = authorizationStructure;
    }

    public List<AssetQuery> createFrom(Iterable<AssetRequestDto> requests) {
        return StreamSupport.stream(requests.spliterator(), false).map(
            assetRequestDto -> assetQueryFor(
                assetRequestDto.name,
                assetRequestDto.kind)).collect(toList());
    }

    private AssetQuery assetQueryFor(String name, AssetKind assetKind) {
        switch (assetKind) {
            case User:
                return new UserQuery(name, authorizationStructure);
            case OrganizationalUnit:
                return new OrganizationalUnitQuery(name, authorizationStructure);
            case Hardware:
                return new HardwareQuery(name, authorizationStructure);
            default:
                throw new IllegalArgumentException();
        }
    }
}
